src/internal/agent/proxy/transform.go

```diff
@@ -0,0 +1,115 @@
+		`(?P<npmPath>(\/(@[\w\.\-\~]+(\/|%2[fF]))?[\w\.\-\~]+(\/-\/([\w\.\-\~]+\/)?[\w\.\-\~]+\.[\w]+)?(\/-rev\/.+)?)|(-\/(npm|v1|user|package)\/.+))$`)
+
+	return transformRegistryPath(baseURL, reqURL, username, npmURLRegex, "npmPath", "npm")
+}
+
+// PipTransformURL finds the pip API path on a given URL and transforms that to align with the offline registry.
+func PipTransformURL(baseURL string, reqURL string, username string) (string, error) {
+	// For further explanation: https://regex101.com/r/lreZiD/1
+	// This regex was created with information from https://github.com/go-gitea/gitea/blob/0e58201d1a8247561809d832eb8f576e05e5d26d/routers/api/packages/api.go#L210
+	pipURLRegex := regexp.MustCompile(`^(?P<proto>[a-z]+:\/\/)(?P<hostPath>.+?)(?P<pipPath>(\/(simple|files\/)[\/\w\-\.\?\=&%#]*?))?$`)
+
+	return transformRegistryPath(baseURL, reqURL, username, pipURLRegex, "pipPath", "pypi")
+}
+
+// GenTransformURL finds the generic API path on a given URL and transforms that to align with the offline registry.
+func GenTransformURL(baseURL string, reqURL string, username string) (string, error) {
+	// For further explanation: https://regex101.com/r/qcg6Gr/1
+	genURLRegex := regexp.MustCompile(`^(?P<proto>[a-z]+:\/\/)(?P<host>.+?)(?P<startPath>\/[\w\-\.%]+?\/[\w\-\.%]+?)?(?P<midPath>\/.+?)??(?P<version>\/[\w\-\.%]+?)?(?P<package>\/[\w\-\.\?\=&%#]+?)$`)
+
+	matches := genURLRegex.FindStringSubmatch(reqURL)
+	idx := genURLRegex.SubexpIndex
+
+	if len(matches) == 0 {
+		// Unable to find a substring match for the regex
+		return "", fmt.Errorf("unable to extract the genericPath from the url %s", reqURL)
+	}
+
+	packageName := matches[idx("startPath")]
+	// NOTE: We remove the protocol and file name so that https://zarf.dev/package/package1.zip and http://zarf.dev/package/package2.zip
+	// resolve to the same "folder" (as they would in real life)
+	sanitizedURL := fmt.Sprintf("%s%s%s", matches[idx("host")], matches[idx("startPath")], matches[idx("midPath")])
+
+	// Add crc32 hash of the url to the end of the package name
+	table := crc32.MakeTable(crc32.IEEE)
+	checksum := crc32.Checksum([]byte(sanitizedURL), table)
+	packageName = fmt.Sprintf("%s-%d", strings.ReplaceAll(packageName, "/", ""), checksum)
+
+	version := matches[idx("version")]
+	if version == "" {
+		version = matches[idx("package")]
+	}
+
+	// TODO: (@WSTARR) %s/api/packages/%s is very Gitea specific but with a config option this could be adapted to GitLab easily
+	// transformedURL := fmt.Sprintf("%s/api/v4/projects/39799997/packages/generic/%s%s%s", baseURL, packageName, version, matches[idx("package")])
+	transformedURL := fmt.Sprintf("%s/api/packages/%s/generic/%s%s%s", baseURL, username, packageName, version, matches[idx("package")])
+
+	return transformedURL, nil
+}
+
+func transformRegistryPath(baseURL string, reqURL string, username string, regex *regexp.Regexp, pathGroup string, regType string) (string, error) {
```

Comment:
I know this is a private func but could we add a docstring to this? At quick glances I'm unsure what some of the params are for (like `pathGroup` and `regType`)

Reply:
Added a comment